File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
dxs = [1, 0, -1, 0]
dys = [0, 1, 0, -1]

def is_reachable(x, y):
    return 0 <= x < m and 0 <= y < n and matrix[x][y] != 1 and not visited[x][y]

def dfs(x, y):
    stack = []

    stack.append([x, y])
    visited[x][y] = True

    cnt = 1

    while stack:
        x_, y_ = stack.pop()

        for dx, dy in zip(dxs, dys):
            nx, ny = x_ + dx, y_ + dy

            if is_reachable(nx, ny):
                cnt += 1
                visited[nx][ny] = True
                stack.append([nx, ny])

    return cnt


m, n, k = map(int, input().split())

matrix = [[0] * n for _ in range(m)]
visited = [[False] * n for _ in range(m)]

for _ in range(k):
    sx, sy, ex, ey = map(int, input().split())

    for y in range(sx, ex):
        for x in range(sy, ey):
            matrix[x][y] = 1

result = []

for x in range(m):
    for y in range(n):
        if not visited[x][y] and matrix[x][y] != 1:
            cnt = dfs(x, y)
            result.append(cnt)

result.sort()

print(len(result))
for res in result:
    print(res, end = " ")

File: 백준/Silver/2583. 영역 구하기/영역 구하기.cc
#include <bits/stdc++.h>
using namespace std;

int n, m, k, i, j, x, y, cnt, area;
int sx, sy, ex, ey;

int arr[101][101];
bool visited[101][101];

const int dxs[4] = {1, 0, -1, 0};
const int dys[4] = {0, 1, 0, -1};

bool is_reachable(int x, int y) {
	return (0 <= x && x < m) && (0 <= y && y < n);
}

int bfs(int x, int y) {
	queue<pair<int, int>> q;

	int count = 0;

	visited[x][y] = true;
	q.push({x, y});

	while (q.size()) {
		tie(x, y) = q.front(); q.pop();
		count++;

		for (i = 0; i < 4; ++i) {
			int nx = x + dxs[i];
			int ny = y + dys[i];

			if (is_reachable(nx, ny) && (arr[nx][ny] == 0) && !visited[nx][ny]) {
				visited[nx][ny] = true;
				q.push({nx, ny});
			}
		}
	}

	return count;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

	cin >> n >> m >> k;

	for (i = 0; i < k; ++i) {
		cin >> sx >> sy >> ex >> ey;

		for (x = sx; x < ex; ++x) {
			for (y = sy; y < ey; ++y) {
				arr[x][y] = 1;
			}
		}
	}

	vector<int> vec;

	for (x = 0; x < m; ++x) {
		for (y = 0; y < n; ++y) {
			if (!visited[x][y] && arr[x][y] == 0) {
				area = bfs(x, y);
				vec.push_back(area);
				cnt += 1;
			}
		}
	}

	sort(vec.begin(), vec.end());

	cout << cnt << "\n";

	for (auto v : vec) {
		cout << v << " ";
	}

	return 0;
}